### PR TITLE
Python 3.6 and  PyQt 5 compatibility

### PR DESCRIPTION
**Pull request by wkitka**
Update lib-maxiv-svgsynoptic to be compatible with Python 3.6, Taurus 4.7, PyQt 5 and drop support for Python 2.

### CODE REVIEW

**Comment by dschick** [comment]
many thanks for putting this forward!
It actually works like a charm for me.

However, when running the examples I need to rename the `tango` folder to something else, because the modules try to import this folder instead of the correct `PyTango` module.

After this change, both examples seem to work as expected.
I was wondering, if a `TaurusDevicePanel` should open, when clicking onto one of the rectangles in the `tango` examples, which does not happens for me.

Best

Daniel

**Comment by wkitka** [comment]
Hej @dschick, 

nice to hear it works for you! 
Yes, `TaurusDevicePanel` should appear, but I'm not sure if for every one of the rectangles. I tested it under CentOS 7 and Ubuntu 18.04 and `TaurusDevicePanel ` appeared.

**Comment by HEnquist** [comment]
I believe only the rectangle in the center of the tango example should give a device panel or right-click. 
Renaming the tango example is probably a good idea, to avoid the import problem.

**Comment by HEnquist** [comment]
Isn't this change large enough to bump to 4.0.0?

File: svgsynoptic2/ttldict.py [hunk 1]
```diff
@@ -65,13 +73,13 @@ def is_expired(self, key, now=None, remove=False):
 
     def __len__(self):
         with self._lock:
-            for key in self._values.keys():
+            for key in list(self._values.keys()):
```
Comment (HEnquist) [inline svgsynoptic2/ttldict.py:76]: skip list conversion

File: svgsynoptic2/ttldict.py [hunk 2]
```diff
@@ -65,13 +73,13 @@ def is_expired(self, key, now=None, remove=False):
...
                 self.is_expired(key, remove=True)
             return len(self._values)
 
     def __iter__(self):
         with self._lock:
-            for key in self._values.keys():
+            for key in list(self._values.keys()):
```
Comment (HEnquist) [inline svgsynoptic2/ttldict.py:82]: skip list conversion

File: svgsynoptic2/taurussynopticwidget.py [hunk 3]
```diff
@@ -364,14 +374,14 @@ def closeEvent(self, event):
         # Get rid of all opened panels, otherwise the application will
         # not exit cleanly.
         super(TaurusSynopticWidget, self).closeEvent(event)
-        for model, panel in self._panels.items():
-            print "closing panel for", model
+        for model, panel in list(self._panels.items()):
```
Comment (HEnquist) [inline svgsynoptic2/taurussynopticwidget.py:377]: skip list conversion

File: svgsynoptic2/taurussynopticwidget.py [hunk 4]
```diff
@@ -46,37 +41,41 @@ def run(self):
...
 
 
 def getStateClasses(state=None):
-    "Return a state CSS class configuration"
+    """
+    Return a state CSS class configuration
+    """
     states = dict((("state-%s" % name), s == state)
-                  for name, s in PyTango.DevState.names.items())
+                  for name, s in list(tango.DevState.names.items()))
```
Comment (HEnquist) [inline svgsynoptic2/taurussynopticwidget.py:53]: skip list conversion

File: svgsynoptic2/taurusregistry.py [hunk 5]
```diff
@@ -160,10 +157,12 @@ def _remove_listener(self, model):
...
 
     def get_listener(self, model):
-        "return the listener for a given model"
+        """
+        return the listener for a given model
+        """
         if model in self.listeners:
             return self.listeners[model]
-        for attr in self.listeners.values():
+        for attr in list(self.listeners.values()):
```
Comment (HEnquist) [inline svgsynoptic2/taurusregistry.py:165]: skip list conversion

File: svgsynoptic2/taurusregistry.py [hunk 6]
```diff
@@ -108,10 +104,11 @@ def handle_event(self, evt_src, *args):
...
 
     def _update(self, attributes=CaselessDict()):
+        """
+        Update the subscriptions; add new ones, remove old ones
+        """
 
-        "Update the subscriptions; add new ones, remove old ones"
-
-        listeners = set(k.lower() for k in self.listeners.keys())
+        listeners = set(k.lower() for k in list(self.listeners.keys()))
```
Comment (HEnquist) [inline svgsynoptic2/taurusregistry.py:111]: skip list conversion

File: svgsynoptic2/taurusregistry.py [hunk 7]
```diff
@@ -65,31 +59,33 @@ def run(self):
...
+        """
         attrs = CaselessDict()
         taurusattrs = self._taurus_attributes
         for model in models:
             if self.device_validator.isValid(model):
                 # for convenience, we subscribe to State for any devices
                 modelstate = model + "/State"
                 attrs[modelstate] = True
-                if modelstate not in taurusattrs.keys():
+                if modelstate not in list(taurusattrs.keys()):
```
Comment (HEnquist) [inline svgsynoptic2/taurusregistry.py:72]: skip list conversion

File: svgsynoptic2/synopticwidget.py [hunk 8]
```diff
@@ -121,41 +118,19 @@ def addBackend():
...
-                    texta,textb = text.split("/*configplaceholder*/", 1)
-                    text =texta + self._modelNames + textb
+                    # print("placeholder found")
+                    texta, textb = text.split("/*configplaceholder*/", 1)
+                    text = texta + self._modelNames + textb
             view.setHtml(text, base_url)
 
         return view
 
-    def _setup_inspector(self, view):
```
Comment (HEnquist) [inline svgsynoptic2/synopticwidget.py:137]: Can the Inspector be opened in another way, or has the functionality been removed in pyqt5?

File: svgsynoptic2/synopticwidget.py [hunk 9]
```diff
@@ -60,44 +61,45 @@ def _setup_ui(self, url=None, section=None):
...
-        #build a javascript defining the models
+    def setConfig(self, config_file):
+        abspath = os.path.dirname(os.path.abspath(config_file))
+        # build a javascript defining the models
         text = "var modelNames ={"
-        with open(configFile, 'r') as read_file:
+        with open(config_file, 'r') as read_file:
             data = json.load(read_file)
-            for key in data.keys():
+            for key in list(data.keys()):
```
Comment (HEnquist) [inline svgsynoptic2/synopticwidget.py:77]: skip list conversion

File: svgsynoptic2/caseless.py [hunk 10]
```diff
@@ -72,12 +72,14 @@ def keys(self):
         return [str(k) for k in self._dict]
 
     def items(self):
-        return zip(self.keys(), self.values())
+        return list(zip(list(self.keys()), list(self.values())))
```
Comment (HEnquist) [inline svgsynoptic2/caseless.py:75]: skip list conversion, should be safe here as well

File: svgsynoptic2/synopticwidget.py (hunk 8 above)
Comment (wkitka) [reply svgsynoptic2/synopticwidget.py:137]: As far as I know there is no equivalent QWebInspector for QWebEngine. QWebEngine replaced QtWebKit in PyQt5.

**Review by johanfforsberg — COMMENTED** [review]
Looks like a clean update! I tried it on my machine and it seems to work fine here too.

We could take the opportunity to remove most of the print statements, at least any that aren't related to error handling. Otherwise :+1:!

**Pull request merged**